core/commands/repo.go

```diff
@@ -406,3 +413,190 @@ var repoVersionCmd = &oldcmds.Command{
+	> ipfs repo has -r QmeUKqHMr5r6wJzt3AVScnUAyC4TkdZGjYJNaFBCL1BmAv
+	QmeUKqHMr5r6wJzt3AVScnUAyC4TkdZGjYJNaFBCL1BmAv	complete	3.2 MB of 3.2 MB (100.00%)
+`,
+	},
+	Arguments: []cmdkit.Argument{
+		cmdkit.StringArg("key", true, true, "Key(s) to check for locality.").EnableStdin(),
+	},
+	Options: []cmdkit.Option{
+		cmdkit.BoolOption("recursive", "r", "Check recursively the graph of objects to build more precise stats"),
+	},
+	Run: func(req cmds.Request, res cmds.ResponseEmitter) {
+		n, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmdkit.ErrNormal)
+			return
+		}
+
+		recursive, _, err := req.Option("recursive").Bool()
+		if err != nil {
+			res.SetError(err, cmdkit.ErrNormal)
+			return
+		}
+
+		// Decode all the keys
+		var keys []*cid.Cid
+		for _, arg := range req.Arguments() {
+			c, err := cid.Decode(arg)
+			if err != nil {
+				res.SetError(err, cmdkit.ErrNormal)
+				return
+			}
+
+			keys = append(keys, c)
+		}
+
+		// an offline DAGService will not fetch from the network
+		offlineDag := dag.NewDAGService(bservice.New(
+			n.Blockstore,
+			offline.Exchange(n.Blockstore),
+		))
+
+		ctx := n.Context()
+
+		for _, k := range keys {
+			root, err := offlineDag.Get(ctx, k)
+			if err != nil && err != dag.ErrNotFound {
+				res.SetError(err, cmdkit.ErrNormal)
+				return
+			}
+
+			hasRoot := err != dag.ErrNotFound
+
+			if !hasRoot || !recursive {
+				// we don't have precise size information
+				res.Emit(&localityOutput{Hash: k.String(), Local: hasRoot})
+				continue
+			}
+
+			local, sizeLocal, sizeTotal, err := walkBlock(ctx, offlineDag, root)
+			if err != nil {
+				res.SetError(err, cmdkit.ErrNormal)
+				return
+			}
+
+			res.Emit(&localityOutput{
+				Hash:      k.String(),
+				Local:     local,
+				SizeLocal: sizeLocal,
+				SizeTotal: sizeTotal,
+			})
+		}
+	},
+	Encoders: cmds.EncoderMap{
+		cmds.Text: cmds.MakeEncoder(func(req cmds.Request, w io.Writer, v interface{}) error {
+			lo, ok := v.(*localityOutput)
+			if !ok {
+				return e.TypeErr(lo, v)
+			}
+
+			recursive, _, err := req.Option("recursive").Bool()
+			if err != nil {
+				return err
+			}
+
+			var state string
+			if recursive {
+				state = "complete"
+				if !lo.Local {
+					state = "incomplete"
+				}
+				if lo.SizeTotal <= 0 {
+					state = "unknown"
+				}
+			} else {
+				if lo.Local {
+					state = "local"
+				} else {
+					state = "non local"
+				}
+			}
+
+			if lo.SizeTotal > 0 {
+				fmt.Fprintf(w, "%s\t%s\t%s of %s (%.2f%%)\n",
+					lo.Hash,
+					state,
+					humanize.Bytes(uint64(lo.SizeLocal)),
+					humanize.Bytes(uint64(lo.SizeTotal)),
+					100.0*float64(lo.SizeLocal)/float64(lo.SizeTotal),
+				)
+			} else {
+				fmt.Fprintf(w, "%s\t%s\n", lo.Hash, state)
+			}
+
+			return nil
+		}),
+	},
+	Type: localityOutput{},
+}
+
+func walkBlock(ctx context.Context, dagserv dag.DAGService, nd node.Node) (bool, int, int, error) {
+	stat, err := nd.Stat()
```

Comment:
So... time to air a dirty secret... This only exists for unixfs. Don't use it. You *can't* know how much of a dag is missing (only how much of a *unixfs* filesystem is missing).

Have you considered using the `ipfs files` feature and extending the `ls` command? I think this functionality may fit better there. Also, MFS (`ipfs files`) is a *much* nicer way to work with IPFS.

Example:

```
> ipfs files cp /ipfs/QmId /some/path
> ipfs files ls --local --tree --size /some/path
/some/path
└── (1MiB/2MiB) foo
    ├── (2MiB) bar.go
    └── (2MiB) baz
```

Reply:
Yeah, i'm on-board with this functionality being in `ipfs files`. Especially since we should be able to do: `ipfs files ls --local --tree --size /ipfs/QmFooBar` (skipping the copy step)

Reply:
(really, we should merge `ipfs files {ls,etc.}` and `ipfs {ls,}`.

> Especially since we should be able to do...

I agree but we can't currently. Is there an issue for this?

Reply:
> So... time to air a dirty secret... This only exists for unixfs. Don't use it. You can't know how much of a dag is missing (only how much of a unixfs filesystem is missing).

I'm confused. If I'm looking at the code of `Stat()`, I don't see anything specific to unixfs. This does works as well:
```
$ echo 'test' | ipfs add
added QmeomffUNfmQy76CQGy9NdmqEnnHU9soCexBnGU3ezPHVH QmeomffUNfmQy76CQGy9NdmqEnnHU9soCexBnGU3ezPHVH
$ ipfs object stat QmeomffUNfmQy76CQGy9NdmqEnnHU9soCexBnGU3ezPHVH
NumLinks: 0
BlockSize: 13
LinksSize: 2
DataSize: 11
CumulativeSize: 13
```
What am I missing ?
Do you mean that `Stat()` is not implemented properly for every kind of ipld block ? If so, why ?

Reply:
More questions:
- `ipfs files` is supposed to be MFS, right ? So if `ipfs repo has` is moved to `ipfs files ls --local --tree --size`, it will only works inside MFS. I'm not using mfs for my project so it's not helpful for me.
- what would the `--size` parameter really mean ? Using it alone would intuitively means 'give me various size information you have'. But adding `--local` would somehow change it's meaning to 'check that everything is local properly and report how much if not' ?

Moving the feature to 'ipfs files ls' feels like shoe-horning to me. I think there is space for a proper 'node administrator' command to inspect what is inside a repository (and in my case help me drive ipfs properly).

Reply:
> Do you mean that Stat() is not implemented properly for every kind of ipld block ? If so, why ?

Yes. To be precise, `Stat`, etc. only apply to DagPB IPLD nodes (so, technically not *only* unixfs). One thing on my todo list is to fix the go interfaces and get rid of that method.

Why? IPLD came after unixfs and the merkledag (DagPB). We wanted to generalize the concept and create a system that:

1. Allowed *structured* linked data. DagPB does not allow much structure. We wanted something more like JSON.
2. Allowed us to interpret and treat all existing existing merkledags (like Ethereum and git) as one big merkleforest. That way, we could write one set of tools to work with any merkle-linked data.

That means we couldn't require any metadata fields like `Size`. We wouldn't have, e.g., been able to interpret a git repo as an IPLD dag if we did that.

---

> I'm not using mfs for my project so it's not helpful for me.

Aside: IMO, you should. It makes managing files much simpler. Eventually, we'd like to using MFS for *all* local file management (where `/ipfs` and `/ipns` would be "virtual" directories within the MFS tree. Eventually, `ipfs CMD` will probably be aliases for `ipfs files CMD` (possibly with some path restrictions).

However, for now, it *would* probably be simpler to just add the flags to the `ipfs ls` command (we can work on unifying the commands later). That should work for your purpose.

> what would the --size parameter really mean ? Using it alone would intuitively means 'give me various size information you have'. But adding --local would somehow change it's meaning to 'check that everything is local properly and report how much if not' ?

Sorry, that was poorly explained.

* `--size` would mean "report the size of the file and how much has been downloaded so far". We could also just add a separate `--local-size` flag for that.
* `--local` would mean "don't fetch anything when running this command". Ideally, we'd provide that as an argument to *all* commands. Otherwise, `ls --tree` would block until all directories (but not their files) have been fetched.

> I think there is space for a proper 'node administrator' command to inspect what is inside a repository (and in my case help me drive ipfs properly).

It would be useful to have a command to check and see if we have locally something without fetching it. However, we *can't* make a general purpose command for getting size information about arbitrary IPLD nodes.

Reply:
Thank you, that's helpful.

> Aside: IMO, you should. It makes managing files much simpler. Eventually, we'd like to using MFS for all local file management (where /ipfs and /ipns would be "virtual" directories within the MFS tree. Eventually, ipfs CMD will probably be aliases for ipfs files CMD (possibly with some path restrictions).

For some context, my project is https://github.com/MichaelMure/Arbore. It's a social file sharing app. For a broader view on how I use IPFS, see https://github.com/ipfs/go-ipfs/issues/3866

Unless I miss something (entirely possible), I see MFS as more a problem than a help for me. I'm using the IPFS repository as a black box / buffer area. In particular, I need to:
- import easily files/directory (`ipfs add -r`). I get a nice progress from the API. I don't understand how to do that with MFS. On top of that I would need to manage path inside of MFS where I really don't need to.
- be able to trigger a download (`ipfs pin add`). MFS doesn't seems to support that (?). Each share in Arbore has it's own root CID, so it's very easy to manage. If I want to remove a share from IPFS, I just `pin rm` the root, trigger a garbage collect and IPFS will deal with everything.
- be able to use the filestore. Not using twice the disk size when you handle big files is a nice touch. I don't think MFS is able to do that.
- be able to encrypt files. Ideally it will be done by IPFS when encryption support is added. Will that be supported by MFS ?
---
> It would be useful to have a command to check and see if we have locally something without fetching it. However, we can't make a general purpose command for getting size information about arbitrary IPLD nodes.

What about a command that return detailed size information only if the nodes are compatible ? If the cumulative size is not available, we are still able to tell if everything is local or not, and how much data is here (using `Node.Size()` recursively). 

---

So, unless I got it wrong, MFS is not an option for me. I see two viable options for me:
- move the feature to `ipfs ls` 
- fix the current PR for `ipfs repo has` to not use `Node.Stat()`

@Stebalien @whyrusleeping Thoughts ?

Reply:
@MichaelMure 

> - import easily files/directory (`ipfs add -r`). I get a nice progress from the API. I don't understand how to do that with MFS. On top of that I would need to manage path inside of MFS where I really don't need to.

Keep using `ipfs add`, Its really supposed to just be an alias for `ipfs files add` but that command doesnt exist yet (see the javascript api).

> - be able to trigger a download (`ipfs pin add`). MFS doesn't seems to support that (?). Each share in Arbore has it's own root CID, so it's very easy to manage. If I want to remove a share from IPFS, I just `pin rm` the root, trigger a garbage collect and IPFS will deal with everything.

For now, keep that flow. We need to add a command to `ipfs files` that lets you fully sync a given bit of content. Like `ipfs files cp —full-sync /ipfs/QmFooBar /path/to/thing` 

> - be able to use the filestore. Not using twice the disk size when you handle big files is a nice touch. I don't think MFS is able to do that.

Keep using the filestore, ipfs files doesnt care where the blocks are.

> - be able to encrypt files. Ideally it will be done by IPFS when encryption support is added. Will that be supported by MFS ?

Yeah, It really won't get in your way at all for any of these concerns.



To be clear, what we're recommending (and as I said in my previous comment) the command would look like:

`ipfs files ls —local —tree —size /ipfs/QmFooBar`

Note that the content in question is *not* explicitly within the mfs namespace.